src/tasks/publish/index.ts

```diff
@@ -0,0 +1,226 @@
+    contractAddress = await _deployMainContract(contractName, noVerify, bre)
+  } else {
+    logMain('Reusing contract from previous version')
+    contractAddress = prevVersion.contractAddress
+  }
+  logMain(`contractAddress: ${contractAddress}`)
+
+  // Prepare release directory
+  // npm run build must create: index.html, src.js, script.js
+  logMain(`Building app front-end at ${appSrcPath}`)
+  await execa('npm', ['run', 'build'], { cwd: appSrcPath })
+
+  // Generate and validate Aragon artifacts, release files
+  logMain(`Generating Aragon app artifacts`)
+  await generateArtifacts(distPath, bre)
+  validateArtifacts(distPath)
+
+  // Upload release directory to IPFS
+  logMain('Uploading release assets to IPFS...')
+  const contentHash = await uploadReleaseToIpfs(distPath, {
+    ipfsProvider,
+    ignorePatterns: ipfsIgnore
+  })
+  logMain(`Release assets uploaded to IPFS: ${contentHash}`)
+
+  // Generate tx to publish new app to aragonPM
+  const versionInfo = {
+    version: nextVersion,
+    contractAddress,
+    contentUri: apm.toContentUri('ipfs', contentHash)
+  }
+
+  const txData = await apm.publishVersion(appName, versionInfo, provider, {
+    managerAddress
+  })
+
+  logMain(
+    `Successfully generated TX data for publishing ${appName} version ${nextVersion}
+
+to: ${txData.to}
+data: ${apm.encodePublishVersionTxData(txData)}
+`
+  )
+}
+
+/**
+ * Returns latest version given a partial or full appName
+ * Returns undefined if repo does not exists
+ * @param appName "finance" | "finance.aragonpm.eth"
+ * @param provider
+ */
+async function _getLastestVersionIfExists(
+  appName: string,
+  provider: ethers.providers.Provider
+): Promise<apm.ApmVersion | undefined> {
+  const fullAppName = getFullAppName(appName)
+  // Check ENS name first since ethers causes an UnhandledPromiseRejectionWarning
+  const repoAddress = await provider.resolveName(fullAppName)
+  if (!repoAddress) return
+
+  // Check for latest version but expect errors
+  try {
+    return await apm.getRepoVersion(repoAddress, 'latest', provider)
+  } catch (e) {
+    if (e.message.includes('ENS name not configured')) return
+    throw e
+  }
+}
+
+/**
+ * Deploys a new implementation contract and returns its address
+ * @param contractName
+ * @param noVerify
+ * @param bre
+ */
+async function _deployMainContract(
+  contractName: string,
+  noVerify: boolean,
+  bre: BuidlerRuntimeEnvironment
+): Promise<string> {
+  // Compile contracts
+  await bre.run(TASK_COMPILE)
+  // Deploy contract
+  const MainContract = bre.artifacts.require(contractName)
+  const mainContract = await MainContract.new()
+  logMain('Implementation contract deployed')
+  const chainId = await bre.web3.eth.getChainId()
+  if (!noVerify && etherscanSupportedChainIds.has(chainId)) {
+    logMain('Verifying on Etherscan')
+    await bre.run(TASK_VERIFY_CONTRACT, {
+      contractName,
+      address: mainContract.address
+      // TODO: constructorArguments
```

Comment:
We can assume for now that we don't need any constructor arguments; AragonApps being published through this process should not be deployed with constructors.